wavemon.1
.TH wavemon 1 "February 2026" Linux "User Manuals"
.SH NAME
wavemon \- a wireless network monitor
.SH SYNOPSIS
.B wavemon [-h] [-i
.I ifname
.B ] [-g] [-v]
.SH DESCRIPTION
\fIwavemon\fR is a ncurses-based monitoring application for wireless network
devices. It plots levels in real-time as well as showing wireless and network
related device information.

The \fIwavemon\fR interface splits into different "screens".
Each screen presents information in a specific manner. For example, the
"info" screen shows current levels as bargraphs, whereas the "level" screen
represents them along a moving timeline.

On startup, you'll see (depending on configuration) one of the different monitor
screens. At the bottom, you'll find a \fImenu-bar\fR listing the screens and
their activating keys. Each screen is activated by either the corresponding
function key (F1..10), its numeric shortcut (1..0), or the underlined shortcut
letter of the screen name.  The following \fIscreens\fR can be selected:
.TP
.B Info (F1 or 'i')
This is the most comprehensive screen. It displays a condensed overview of
wireless-specific parameters and network statistics, as well as bar graphs.
The layout is arranged into several sub-sections.

The
.B Interface
section at the top shows information about the monitoring interface,
including interface name, type, ESSID, and available encryption formats.

Below, in the
.B Levels
section, you can see up to four bargraphs showing (1) relative signal quality
and (2) signal level in dBm. If the wireless driver also supports noise level
information, additionally (3) noise level in dBm and (4) Signal-Noise-Ratio
(SNR) in dB are shown.  The colour of the signal level bargraph changes
from red to yellow and green at fixed levels. If thresholds have been set,
two arrows on the signal level graph will show the positions of the current
thresholds.

The
.B Statistics
section displays packet and byte counters and a few other packet-related
statistics.

The subsequent
.B Info
subsection lists the current operational mode and configuration of the
wireless interface. What parameters are actually shown depends on the
capabilities and selected mode of your network device.

Lastly, the
.B Network
section shows network-level parameters. The MAC-address is resolved from
\fBethers\fR(5). The IPv4 address is shown in CIDR notation (RFC\ 4632
\fIaddress\fR\fB/\fR\fIprefix_len\fR format). Since often those two values
also determine the broadcast address (last 32 \- \fIprefix_len\fR bits set
to 1), that address is shown only if it does not derive from the interface
address and prefix length. Likewise, the interface MTU is shown only if it
differs from the default Ethernet MTU of 1500 bytes.
.TP
.B Level history (F2 or 'l')
This is a full-screen history plot showing the evolution of the signal
level over time.
The screen is partitioned into a grid, with dBm levels shown in green at
the right hand side.
.TP
.B Scan window (F3 or 's')
A periodically updated network scan, showing access points and other
wireless clients. It is sorted depending on \fIsort_order\fR and \fIsort_ascending\fR, see \fBwavemonrc\fR(5).
Each entry starts with the ESSID, followed by the colour-coded MAC
address and the signal/channel information. A green/red MAC address indicates
an (un-)encrypted access point, the colour changes to yellow for non-access
points (in this case the mode is shown at the end of the line). The
uncoloured information following the MAC address lists relative and
absolute signal strengths, channel, frequency, and station-specific information.
The station-specific information includes the station type (ESS for Access Point,
IBSS for Ad-Hoc network), station count and channel utilisation.

A status line at the bottom informs about the current sort order and a few
statistics, such as most (least) crowded channels (least crowded channels
are listed when sorting by descending channel).

The \fIsort_order\fR can also directly be changed via these keyboard shortcuts:
\fIa\fRscending, \fId\fRescending; by \fIe\fRssid, \fIs\fRignal, \fIc\fRhannel (\fIC\fR also with signal),
\fIm\fRac address, or by \fIo\fRpen access (\fIO\fR also with signal).

Please note that sorting order changes at the time new data comes in,
not when the setting is activated.

You can \fIfilter\fR the bands via these keyboard shortcuts: \fI2\fR (2.4GHz only),
\fI5\fR (5GHz only), and \fIb\fR (both bands). Hidden ESSIDs can be excluded from
display via the \fIh\fR shortcut.

.TP
.B Preferences (F7 or 'p')
This screen allows you to change all program options such as interface and
level scale parameters, and to save the new settings to the configuration
file. Select a parameter with <up> and <down>, then change the value with
<left> and <right>. Please refer to \fBwavemonrc\fR(5) for an in-depth
description of applicable settings.
.TP
.B Help (F8 or 'h')
This page might show an online-help.
.TP
.B About (F9 or 'a')
Release information and contact URLs.
.TP
\fBQuit (F10 or 'q')
Exit \fIwavemon\fR.
.LP

.SH "OPTIONS"
.IP "\fB\-i \fIinterface\fR\fR"
override autodetection and use the specified \fIinterface\fR.
.IP "\fB\-g\fR"
check screen \fIgeometry\fR: a minimum size is required for proper display; this flag
adds a check to ensure it is sufficiently large. Enable this if window does not display
properly.
.IP "\fB\-h\fR"
print help and exit.
.IP "\fB\-v\fR"
print version information and exit.
.SH Troubleshooting
.IP \(bu
\fIwavemon\fR will exit with \fB'no supported wireless interfaces found'\fR if no usable wireless interfaces
were detected. Check if your wireless interface is otherwise usable, using e.g. \fBiw\fR(8).
If your interface is not listed, causes can be a missing (or not loaded) kernel module; or missing firmware,
which some cards need to operate.

.IP \(bu
Some operations, such as displaying encryption information or performing scans, require
\fBCAP_NET_ADMIN\fR privileges (see \fBcapabilities\fR(7)). For non-root users, these can be
enabled by installing \fIwavemon\fR setuid-root.

.IP \(bu
Running \fIwavemon\fR inside \fBscreen\fR(1) may cause garbled output on certain locales.
This can be fixed calling screen like this:
.EX
$ LC_ALL=C screen
.EE
.SH FILES
.IP $XDG_CONFIG_HOME/wavemon/wavemonrc
The local per-user configuration file.
.SH "AUTHOR"
Written by Jan Morgenstern <jan@jm-music.de>.
.SH "REPORTING BUGS"
Open an issue on \fIhttps://github.com/uoaerg/wavemon/issues\fR.
.SH "SEE ALSO"
.BR wavemonrc (5),
.BR ethers (5),
.BR capabilities (7),
.BR iw (8),
.BR locale (1)


NEWS
NEWS
====

--------------------
0.7.6 (2014-01-18)
--------------------
  * Scan window:
    - complete redesign based on pthreads to fix terminal bug, as a result of better
      design the scan window now also can be suspended (details of fix are below),
    - add message for 'too much data' (E2BIG) error case,
    - use same formatting (3 digits) for 2/5 GHz,
    - support for changing sort order via keyboard shortcuts (see manpage, several
      ascending/descending sort orders are now available directly on the screen),
    - currently active sort order is now displayed on status line at the bottom,
    - stabilized by-channel sort to always use the same order for identical channels,
    - added 'sort by Essid' to selection of sort menus;
  * Info window:
    - add percentage for easier reading of link quality,
    - deprecate use of locale settings for 'discard' information (fixes display);
  * Configuration screen:
    - add support for Page Up/Down, Home, and End buttons to ease navigation,
    - added more information to '-v' version information (ncurses/WE versions);
  * Bug fixes:
    - fix bug in ranking top/bottom 3 channels in scan window,
    - complete redesign of scan process based on pthreads to fix bug that became
       visible on ncurses >= 5.9.20130504 (slackware 14.1 and debian jessie), caused
       by child/parent both updating the same screen (bug had been hidden by the
       internal buffering of ncurses which was then removed), reported by Nick Warne.

--------------------
0.7.5 (2012-05-04)
--------------------
  * Scan window:
    - added a summary status line to present summary statistics.
  * Bug fixes:
    - added additional keyboard shortcuts for each screen to cope with
      terminals which remap the function keys, fixing Ubuntu bug #725036;
    - the shortcut is the first character of the screen's name (F1: i,
      F2: l, F3: s, F7: p, F8: h, F9: a, F10: q), see manpage for details;
    - added work-around for terminals with broken terminfo entries (e.g.
      old aterm), now supporting the vt100 PF1..4 sequences in addition.

--------------------
0.7.4 (2012-02-03)
--------------------
  * General:
    - deprecate reliance on /proc/net/wireless, since tests on Tiny Core Linux
      showed that it no longer reliably lists available wireless interfaces,
    - instead use safe route of using /proc/net/dev, probing each interface
      in turn for wireless extensions;
  * Scan window:
    - configurable scan sort order (at runtime and via configuration file),
    - can sort by (combinations of) channel, signal strength, and openness,
    - visually distinguish 2.4GHz and 5.8GHz channels;
  * Fixes:
    - wavemonrc.5 manpage fixes thanks to Jonathan McCrohan,
    - support common options (-v/-h) even if no interfaces found,
    - but do not start operation unless at least 1 wireless interface exists.

--------------------
0.7.3 (2011-12-22)
--------------------
  * Scan window:
    - items are now sorted first by channel (ascending) then by signal
      strength (descending), as before
    - using one line per entry doubled the capacity of the scan window
    - refresh interval is made configurable (via 'Statistics updates')
  * Configuration screen:
    - memorizes the last-changed item to ease longer configuration tasks
  * Fixes:
    - better (cleaner) termination on error
    - fixed clear-line refresh in info screen
    - fixed bug in generator for random link quality (causing negative values)
    - reverted using separate process group (problems when running under sudo)
    - miscellaneous code reordering to decouple functional units

--------------------
0.7.2 (2011-02-25)
--------------------
  * General:
    - support SIGWINCH for resizing (thanks to Koniu)
    - reorganized screen handling to support resizing
    - make geometry checking optional and off by default
    - allow suspending wavemon (CTRL-Z) on most screens
  * Info screen:
    - suppress obvious broadcast address (deriving from prefix)
    - special case for "no interface address assigned"
    - display interface MTU if different from the default value
    - show interface flags and TX queue length if enough space
  * Scan window:
    - set up enable interface if it is down (otherwise no scan)
    - drop support for wireless extensions < 18 (minimum requirement)
    - clean up child scan process on error (avoid garbled screen)
  * Configuration screen:
    - support 'Cisco-style' MAC addresses (dots rather than colons)
  * Fixes:
    - do not report zero bitrates (Info screen, 802.11n interface)
    - remove unused timer in Scan window (data sampling is now
      confined to info/histogram screens)
    - fixed various compiler warnings
    - fixed several problems in and improved the random generator
    - use sigaction() rather than signal()
    - miscellaneous reorganization of the configuration menu code

--------------------
0.7.1 (2010-12-24)
--------------------
  * Scan window:
    - improved and more consistent formatting of entries
    - better responsiveness by running scan operation as child process
  * Security capabilities:
    - display capabilities supported by the driver
    - display current and available key sizes
    - decode relevant IE information
  * check and warn about insufficient permissions (CAP_NET_ADMIN) that
    are required for several wireless operations (e.g. scanning)
  * Bug Fixes:
    - avoid triggering assertion on large screens
    - require minimum version (18) of the wireless extensions
    - better clearing of info screen
    - various cleanups
  * overhaul of manpages

--------------------
0.7.0 (2010-10-18)
--------------------
  * NEW Scan Window: this replaces the old access point screen. It displays
    a sorted list (in descending order of signal quality) of access points
    and other wireless clients within range, periodically refreshed at a
    configurable rate.
  * Access Point screen: finally deprecated this - the SIOCGIWAPLIST

wavemon.c
/*
 * wavemon - a wireless network monitoring application
 *
 * Copyright (c) 2001-2002 Jan Morgenstern <jan@jm-music.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
#include "wavemon.h"
#include <locale.h>
#include <setjmp.h>

/* GLOBALS */

/**
 * screen switching table
 * @key_name:	name under which the screen appears in the menu bar
 * @shortcut:   1-character shortcut key for @key_name
 * @init:	screen initialisation function pointer
 * @loop:	screen update function pointer (connected to menu)
 * @fini:	screen cleanup function pointer
 */
static const struct {
	const char *const	key_name;
	const char              shortcut;
	void		 	(*init)(void);
	int			(*loop)(WINDOW *);
	void			(*fini)(void);
} screens[] = {
	[SCR_INFO]	= {
		.key_name = "info",
		.shortcut = 'i',
		.init	  = scr_info_init,
		.loop	  = scr_info_loop,
		.fini	  = scr_info_fini
	},
	[SCR_LHIST]	= {
		.key_name = "lhist",
		.shortcut = 'l',
		.init	  = scr_lhist_init,
		.loop	  = scr_lhist_loop,
		.fini	  = scr_lhist_fini
	},
	[SCR_SCAN]	= {
		.key_name = "scan",
		.shortcut = 's',
		.init	  = scr_aplst_init,
		.loop	  = scr_aplst_loop,
		.fini	  = scr_aplst_fini
	},
	[SCR_EMPTY_F4]	= {
		.key_name = "",
	},
	[SCR_EMPTY_F5]	= {
		.key_name = "",
	},
	[SCR_EMPTY_F6]	= {
		.key_name = "",
	},
	[SCR_PREFS]	= {
		.key_name = "prefs",
		.shortcut = 'p',
		.init	  = scr_conf_init,
		.loop	  = scr_conf_loop,
		.fini	  = scr_conf_fini
	},
	[SCR_HELP]	= {
		.key_name = "help",
		.shortcut = 'h',
		.init	  = scr_help_init,
		.loop	  = scr_help_loop,
		.fini	  = scr_help_fini
	},
	[SCR_ABOUT]	= {
		.key_name = "about",
		.shortcut = 'a',
		.init	  = scr_about_init,
		.loop	  = scr_about_loop,
		.fini	  = scr_about_fini
	},
	[SCR_QUIT]	= {
		.key_name = "quit",
		.shortcut = 'q',
	}
};

/*
 * SIGWINCH handling with buffer synchronisation variable
 */
static sigjmp_buf		env_winch;
static volatile sig_atomic_t	env_winch_ready;

static void sig_winch(int __attribute__((unused))signo)
{
	if (env_winch_ready) {
		env_winch_ready = false;
		siglongjmp(env_winch, 1);
	}
}

static WINDOW *init_menubar(const enum wavemon_screen active)
{
	WINDOW *menu = newwin(1, WAV_WIDTH, WAV_HEIGHT, 0);
	char fkey[8];

	nodelay(menu, TRUE);
	keypad(menu, TRUE);
	wmove(menu, 0, 0);
	for (enum wavemon_screen cur = SCR_INFO; cur <= SCR_QUIT; cur++) {
		const char *p = screens[cur].key_name;
		const int attrs = cur != active ? COLOR_PAIR(CP_CYAN)
						: COLOR_PAIR(CP_CYAN_ON_BLUE) | A_BOLD;

		if (*p) {
			snprintf(fkey, sizeof(fkey), "F%d", cur + 1);
			wattrset(menu, A_REVERSE | A_BOLD);
			waddstr(menu, fkey);

			wattrset(menu, attrs);

			for (int i = 0; i < MAX_MENU_KEY; i++) {
				if (*p == screens[cur].shortcut)	{
					wattron(menu, A_UNDERLINE);
					waddch(menu, *p++);
					wattroff(menu, A_UNDERLINE);
				} else if (*p) {
					waddch(menu, *p++);
				} else {
					wattroff(menu, attrs);
					waddch(menu, ' ');
				}
			}
		}
	}
	wrefresh(menu);

	return menu;
}

static void check_geometry(void)
{
	if (conf.check_geometry &&
	    (LINES < MIN_SCREEN_LINES || COLS < MIN_SCREEN_COLS))
		err_quit("need at least a screen of %ux%u, have only %ux%u",
			    MIN_SCREEN_LINES, MIN_SCREEN_COLS, LINES, COLS);
}

int main(int argc, char *argv[])
{
	int bg_color = COLOR_BLACK;
	volatile enum wavemon_screen cur, next;
	sigset_t blockmask, oldmask;

	getconf(argc, argv);

	if (!isatty(STDIN_FILENO))
		errx(1, "input is not from a terminal");

	setlocale(LC_ALL, "");

	/* initialize the ncurses interface */
	initscr();
	noecho();
	nonl();
	cbreak();
	curs_set(0);
	clear();
	check_geometry();

	start_color();
	if (conf.transparent_bg) {
		bg_color = -1;
		use_default_colors();
	}

	init_pair(CP_RED,	COLOR_RED,	bg_color);
	init_pair(CP_YELLOW,	COLOR_YELLOW,	bg_color);
	init_pair(CP_GREEN,	COLOR_GREEN,	bg_color);
	init_pair(CP_CYAN,	COLOR_CYAN,	bg_color);
	init_pair(CP_BLUE,	COLOR_BLUE,	bg_color);

	init_pair(CP_RED_ON_BLUE,	COLOR_RED,	COLOR_BLUE);
	init_pair(CP_GREEN_ON_BLUE,	COLOR_GREEN,	COLOR_BLUE);
	init_pair(CP_CYAN_ON_BLUE,	COLOR_CYAN,	COLOR_BLUE);
	init_pair(CP_BLUE_ON_BLUE,	COLOR_BLUE,	COLOR_BLUE);

	/* Override signal handlers installed during ncurses initialisation. */
	xsignal(SIGCHLD, SIG_IGN);
	xsignal(SIGWINCH, sig_winch);	/* triggers only when env_winch_ready */
	sigemptyset(&blockmask);
	sigaddset(&blockmask, SIGWINCH);

	for (cur = conf.startup_scr; cur != SCR_QUIT; cur = next) {
		WINDOW *w_menu;
		volatile int escape = 0;

		if (sigprocmask(SIG_BLOCK, &blockmask, &oldmask) < 0)
			err_sys("cannot block SIGWINCH");

		next = cur;
		w_menu = init_menubar(cur);
		(*screens[cur].init)();

		if (sigprocmask(SIG_SETMASK, &oldmask, NULL) < 0)
			err_sys("cannot unblock SIGWINCH");

		if (sigsetjmp(env_winch, true) == 0) {
			env_winch_ready = true;

			do {
				int key = (*screens[cur].loop)(w_menu);

				if (key <= 0)
					usleep(5000);
				/*
				 * Translate vt100 PF1..4 escape sequences sent
				 * by some X terminals (e.g. aterm) into F1..F4.
				 */
				switch (key) {
				case 033:
					escape = 1;
					break;
				case 'O':
					escape = 2;
					break;
				default:
					if ('P' <= key && key <= 'S' && escape == 2)
						key = KEY_F(key - 'P' + 1);
					escape = 0;
				}

				/* Main menu */
				for (enum wavemon_screen s = SCR_INFO; s <= SCR_QUIT; s++) {
					if (*screens[s].key_name && (
					    (unsigned)key == (s == SCR_QUIT ? '0' : '1' + s) ||
					    (unsigned)key == KEY_F(s + 1) ||
					    key == screens[s].shortcut)) {
						next = s;
						break;
					}
				}
			} while (next == cur);
		}

		delwin(w_menu);
		(*screens[cur].fini)();

		/*
		 * next = cur is set in the protected critical section before
		 * sigsetjmp. Due to the loop condition, it can not occur when
		 * no SIGWINCH occurred, hence it indicates a resizing event.
		 */
		if (next == cur) {
			struct winsize size;

			if (ioctl(STDIN_FILENO, TIOCGWINSZ, &size) < 0)
				err_sys("can not determine terminal size");
			resizeterm(size.ws_row, size.ws_col);
			check_geometry();
		}
		clear();
		refresh();
	}
	curs_set(1);
	endwin();

	return EXIT_SUCCESS;
}
